4.1: test_custom_test and test_minimal fails

Both tests fails similarly:
```
_________________________ RatingsTest.test_custom_test _________________________

self = <pyroma.tests.RatingsTest testMethod=test_custom_test>

    def test_custom_test(self):
        rating = self._get_file_rating("custom_test")

>       self.assertEqual(
            rating,
            (   
                2,
                [   
                    "The package's version number does not comply with PEP-386 or PEP-440.",
                    "The package's description should be longer than 10 characters.",
                    "The package's long_description is quite short.",
                    "Your package does not have classifiers data.",
                    "The classifiers should specify what Python versions you support.",
                    (   
                        "You should specify what Python versions you support with "
                        "the 'requires-python'/'python_requires' metadata."
                    ),
                    "Your package does not have keywords data.",
                    "Your package does not have author data.",
                    "Your package does not have author_email data.",
                    "Your package should have a 'url' field with a link to the project home page, or a "
                    "'project_urls' field, with a dictionary of links, or both.",
                    "Your package does neither have a license field nor any license classifiers.",
                    "Specifying a development status in the classifiers gives users "
                    "a hint of how stable your software is.",
                ],
            ),
        )

pyroma/tests.py:226:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib/python3.9/vendor-packages/teamcity/diff_tools.py:33: in _patched_equals
    old(self, first, second, msg)
E   AssertionError: Tuples differ: (1, ["The package's version number does no[1209 chars]g."]) != (2, ["The package's version number does no[838 chars]s.'])
E
E   First differing element 0:
E   1
E   2
E
E   - (1,
E   + (2,
E      ["The package's version number does not comply with PEP-386 or PEP-440.",
E       "The package's description should be longer than 10 characters.",
E       "The package's long_description is quite short.",
E       'Your package does not have classifiers data.',
E       'The classifiers should specify what Python versions you support.',
E       'You should specify what Python versions you support with the '
E       "'requires-python'/'python_requires' metadata.",
E       'Your package does not have keywords data.',
E       'Your package does not have author data.',
E       'Your package does not have author_email data.',
E       "Your package should have a 'url' field with a link to the project home "
E       "page, or a 'project_urls' field, with a dictionary of links, or both.",
E       'Your package does neither have a license field nor any license classifiers.',
E       'Specifying a development status in the classifiers gives users a hint of '
E   -   'how stable your software is.',
E   ?                                 ^
E
E   +   'how stable your software is.'])
E   ?                                 ^^
E
E   -   'Your Cheese may have spoiled!! The only way to gather metadata from your '
E   -   'package was to execute a patched setup.py. This indicates that your package '
E   -   "is using very old packaging techniques, (or that your setup.py isn't "
E   -   'executable at all), and Pyroma will soon regard that as a complete '
E   -   'failure!\n'
E   -   'Please modernize your packaging! If it is already modern, this is a bug.'])
ERROR    root:projectdata.py:78 Exception raised during metadata preparation
Traceback (most recent call last):
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 466, in _handle_backend
    yield
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 459, in _call_backend
    basename: str = callback(outdir, config_settings, **kwargs)
  File "/usr/lib/python3.9/vendor-packages/pyproject_hooks/_impl.py", line 186, in prepare_metadata_for_build_wheel
    return self._call_hook('prepare_metadata_for_build_wheel', {
  File "/usr/lib/python3.9/vendor-packages/pyproject_hooks/_impl.py", line 311, in _call_hook
    self._subprocess_runner(
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 302, in _runner
    self._hook_runner(cmd, cwd, extra_environ)
  File "/usr/lib/python3.9/vendor-packages/pep517/wrappers.py", line 76, in quiet_subprocess_runner
    check_output(cmd, cwd=cwd, env=env, stderr=STDOUT)
  File "/usr/lib/python3.9/subprocess.py", line 424, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/usr/bin/python3.9', '/usr/lib/python3.9/vendor-packages/pyproject_hooks/_in_process/_in_process.py', 'prepare_metadata_for_build_wheel', '/tmp/tmph_9c9vmi']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "$(BUILD_DIR)/pyroma/projectdata.py", line 60, in get_data
    return get_build_data(path)
  File "$(BUILD_DIR)/pyroma/projectdata.py", line 29, in get_build_data
    metadata_dir = build.ProjectBuilder(str(path), runner=pep517.quiet_subprocess_runner).prepare("wheel", tempdir)
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 384, in prepare
    return self._call_backend(
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 459, in _call_backend
    basename: str = callback(outdir, config_settings, **kwargs)
  File "/usr/lib/python3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 474, in _handle_backend
    raise BuildBackendException(  # noqa: B904 # use raise from
build.BuildBackendException: Backend subprocess exited when trying to invoke prepare_metadata_for_build_wheel
___________________________ RatingsTest.test_minimal ___________________________

self = <pyroma.tests.RatingsTest testMethod=test_minimal>

    def test_minimal(self):
        rating = self._get_file_rating("minimal")

>       self.assertEqual(
            rating,
            (
                2,
                [
                    "The package's version number does not comply with PEP-386 or PEP-440.",
                    "The package's description should be longer than 10 characters.",
                    "The package's long_description is quite short.",
                    "Your package does not have classifiers data.",
                    "The classifiers should specify what Python versions you support.",
                    (
                        "You should specify what Python versions you support with "
                        "the 'requires-python'/'python_requires' metadata."
                    ),
                    "Your package does not have keywords data.",
                    "Your package does not have author data.",
                    "Your package does not have author_email data.",
                    "Your package should have a 'url' field with a link to the project home page, or a "
                    "'project_urls' field, with a dictionary of links, or both.",
                    "Your package does neither have a license field nor any license classifiers.",
                    "Specifying a development status in the classifiers gives users "
                    "a hint of how stable your software is.",
                ],
            ),
        )

pyroma/tests.py:167:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib/python3.9/vendor-packages/teamcity/diff_tools.py:33: in _patched_equals
    old(self, first, second, msg)
E   AssertionError: Tuples differ: (1, ["The package's version number does no[1209 chars]g."]) != (2, ["The package's version number does no[838 chars]s.'])
E
E   First differing element 0:
E   1
E   2
E
E   - (1,
E   + (2,
E      ["The package's version number does not comply with PEP-386 or PEP-440.",
E       "The package's description should be longer than 10 characters.",
E       "The package's long_description is quite short.",
E       'Your package does not have classifiers data.',
E       'The classifiers should specify what Python versions you support.',
E       'You should specify what Python versions you support with the '
E       "'requires-python'/'python_requires' metadata.",
E       'Your package does not have keywords data.',
E       'Your package does not have author data.',
E       'Your package does not have author_email data.',
E       "Your package should have a 'url' field with a link to the project home "
E       "page, or a 'project_urls' field, with a dictionary of links, or both.",
E       'Your package does neither have a license field nor any license classifiers.',
E       'Specifying a development status in the classifiers gives users a hint of '
E   -   'how stable your software is.',
E   ?                                 ^
E
E   +   'how stable your software is.'])
E   ?                                 ^^
E
E   -   'Your Cheese may have spoiled!! The only way to gather metadata from your '
E   -   'package was to execute a patched setup.py. This indicates that your package '
E   -   "is using very old packaging techniques, (or that your setup.py isn't "
E   -   'executable at all), and Pyroma will soon regard that as a complete '
E   -   'failure!\n'
E   -   'Please modernize your packaging! If it is already modern, this is a bug.'])
------------------------------ Captured log call -------------------------------
ERROR    root:projectdata.py:78 Exception raised during metadata preparation
Traceback (most recent call last):
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 466, in _handle_backend
    yield
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 459, in _call_backend
    basename: str = callback(outdir, config_settings, **kwargs)
  File "/usr/lib/python3.9/vendor-packages/pyproject_hooks/_impl.py", line 186, in prepare_metadata_for_build_wheel
    return self._call_hook('prepare_metadata_for_build_wheel', {
  File "/usr/lib/python3.9/vendor-packages/pyproject_hooks/_impl.py", line 311, in _call_hook
    self._subprocess_runner(
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 302, in _runner
    self._hook_runner(cmd, cwd, extra_environ)
  File "/usr/lib/python3.9/vendor-packages/pep517/wrappers.py", line 76, in quiet_subprocess_runner
    check_output(cmd, cwd=cwd, env=env, stderr=STDOUT)
  File "/usr/lib/python3.9/subprocess.py", line 424, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/usr/bin/python3.9', '/usr/lib/python3.9/vendor-packages/pyproject_hooks/_in_process/_in_process.py', 'prepare_metadata_for_build_wheel', '/tmp/tmpkx3b9dta']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "$(BUILD_DIR)/pyroma/projectdata.py", line 60, in get_data
    return get_build_data(path)
  File "$(BUILD_DIR)/pyroma/projectdata.py", line 29, in get_build_data
    metadata_dir = build.ProjectBuilder(str(path), runner=pep517.quiet_subprocess_runner).prepare("wheel", tempdir)
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 384, in prepare
    return self._call_backend(
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 459, in _call_backend
    basename: str = callback(outdir, config_settings, **kwargs)
  File "/usr/lib/python3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/usr/lib/python3.9/vendor-packages/build/__init__.py", line 474, in _handle_backend
    raise BuildBackendException(  # noqa: B904 # use raise from
build.BuildBackendException: Backend subprocess exited when trying to invoke prepare_metadata_for_build_wheel
```
In addition to the above failures these warnings are reported:
```
=============================== warnings summary ===============================
pyroma/tests.py::RatingsTest::test_custom_test
pyroma/tests.py::RatingsTest::test_minimal
  /usr/lib/python3.9/vendor-packages/setuptools/dist.py:548: UserWarning: The version specified ('0.0foo') is an invalid version, this may not work as expected with newer versions of setuptools, pip, and PyPI. Please see PEP 440 for more details.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```
OS: OpenIndiana
setuptools: 67.1.0

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -4,7 +4,18 @@ Changelog
 4.2 (unreleased)
 ----------------
 
-- Nothing changed yet.
+- Fall back to installing project's build backend in an isolated environment
+  if a compatible version isn't installed in the current env [CAM-Gerlach]
+
+- Fix metadata extraction failure when project ``long_description`` is included
+  as a header rather than a payload in the ``METADTA`` file [CAM-Gerlach]
+
+- Add a fallback to restore compatibility with Setuptools <61 [CAM-Gerlach]
+
+- Fix tests failing due to invalid versions on Setuptools >=66 [CAM-Gerlach]
+
+- Add ``python_requires``, update classifiers, add implicit dependencies
+  and remove unused deps in Pyroma's own packaging metadata [CAM-Gerlach]
 
 
 4.1 (2022-11-24)
@@ -25,7 +36,6 @@ Changelog
 - Now follows black and flake8 rules
 
 - Check if author_email field contains author name [bessman]
-
 
 
 4.0 (2022-04-14)

--- a/pyroma/projectdata.py
+++ b/pyroma/projectdata.py
@@ -1,13 +1,9 @@
 # Extracts information from a project that has a distutils setup.py file.
-import build
-import email
-import email.policy
+import build.util
 import logging
 import os
 import pathlib
-import pep517
 import sys
-import tempfile
 import tokenize
 from copy import copy
 from distutils import core
@@ -25,11 +21,7 @@ METADATA_MAP = {
 
 
 def get_build_data(path):
-    with tempfile.TemporaryDirectory() as tempdir:
-        metadata_dir = build.ProjectBuilder(str(path), runner=pep517.quiet_subprocess_runner).prepare("wheel", tempdir)
-        with open(pathlib.Path(metadata_dir) / "METADATA", "rb") as metadata_file:
-            metadata = email.message_from_binary_file(metadata_file, policy=email.policy.compat32)
-
+    metadata = build.util.project_wheel_metadata(path, isolated=True)
     if "Description" not in metadata.keys():
         # Having the description as a payload tends to add two newlines, we clean that up here:
         long_description = metadata.get_payload().strip() + "\n"

--- a/pyroma/projectdata.py
+++ b/pyroma/projectdata.py
@@ -35,10 +35,11 @@ def build_metadata(path, isolated=None):
 
 
 def map_metadata_keys(metadata):
+    data = {}
     if "Description" not in metadata.keys():
         # Having the description as a payload tends to add two newlines, we clean that up here:
         long_description = metadata.get_payload().strip() + "\n"
-        data = {"long_description": long_description}
+        data["long_description"] = long_description
 
     for key in set(metadata.keys()):
         value = metadata.get_all(key)

--- a/pyroma/projectdata.py
+++ b/pyroma/projectdata.py
@@ -1,4 +1,5 @@
 # Extracts information from a project that has a distutils setup.py file.
+import build
 import build.util
 import logging
 import os
@@ -20,8 +21,20 @@ METADATA_MAP = {
 }
 
 
-def get_build_data(path):
-    metadata = build.util.project_wheel_metadata(path, isolated=True)
+def build_metadata(path, isolated=None):
+    # If explictly specified whether to use isolation, pass it directly
+    if isolated is not None:
+        return build.util.project_wheel_metadata(path, isolated=isolated)
+
+    # Otherwise, try without build isolation first for efficiency
+    try:
+        return build.util.project_wheel_metadata(path, isolated=False)
+    # If building with build isolation fails, e.g. missing build deps, try with it
+    except build.BuildBackendException:
+        return build.util.project_wheel_metadata(path, isolated=True)
+
+
+def map_metadata_keys(metadata):
     if "Description" not in metadata.keys():
         # Having the description as a payload tends to add two newlines, we clean that up here:
         long_description = metadata.get_payload().strip() + "\n"
@@ -38,6 +51,11 @@ def get_build_data(path):
             key = METADATA_MAP[key]
         data[key] = value
     return data
+
+
+def get_build_data(path, isolated=None):
+    metadata = build_metadata(path, isolated=isolated)
+    return map_metadata_keys(metadata)
 
 
 def get_setupcfg_data(path):

--- a/pyroma/projectdata.py
+++ b/pyroma/projectdata.py
@@ -8,7 +8,11 @@ import sys
 import tokenize
 from copy import copy
 from distutils import core
-from setuptools import config
+
+try:  # config renamed to config.setupcfg on Setuptools >=61 adding pyproject.toml support
+    from setuptools.config.setupcfg import read_configuration
+except ModuleNotFoundError:
+    from setuptools.config import read_configuration
 
 METADATA_MAP = {
     "summary": "description",
@@ -61,7 +65,7 @@ def get_build_data(path, isolated=None):
 
 def get_setupcfg_data(path):
     # Note: By default, setup.cfg will read the pyroma.git/setup.cfg - forcing explicit setup.cfg under test's file path
-    data = config.setupcfg.read_configuration(str(pathlib.Path(path) / "setup.cfg"))
+    data = read_configuration(str(pathlib.Path(path) / "setup.cfg"))
     metadata = data["metadata"]
     return metadata
 
@@ -256,9 +260,7 @@ def get_setuppy_data(path):
 
             elif os.path.isfile("setup.cfg"):
                 try:
-                    from setuptools import config
-
-                    data = config.read_configuration("setup.cfg")
+                    data = read_configuration("setup.cfg")
                     metadata = data["metadata"]
                     metadata["_setuptools"] = True
                 except Exception as e:

--- a/pyroma/testdata/custom_test/setup.py
+++ b/pyroma/testdata/custom_test/setup.py
@@ -1,7 +1,7 @@
 from setuptools import setup, find_packages
 from setuptools.command.test import test
 
-version = "0.0foo"
+version = "0.0.1"
 
 
 class CustomTest(test):

--- a/pyroma/testdata/minimal/setup.py
+++ b/pyroma/testdata/minimal/setup.py
@@ -1,6 +1,6 @@
 from setuptools import setup, find_packages
 
-version = "0.0foo"
+version = "0.0.1"
 
 setup(
     name="minimal",

--- a/pyroma/tests.py
+++ b/pyroma/tests.py
@@ -169,7 +169,6 @@ class RatingsTest(unittest.TestCase):
             (
                 2,
                 [
-                    "The package's version number does not comply with PEP-386 or PEP-440.",
                     "The package's description should be longer than 10 characters.",
                     "The package's long_description is quite short.",
                     "Your package does not have classifiers data.",
@@ -228,7 +227,6 @@ class RatingsTest(unittest.TestCase):
             (
                 2,
                 [
-                    "The package's version number does not comply with PEP-386 or PEP-440.",
                     "The package's description should be longer than 10 characters.",
                     "The package's long_description is quite short.",
                     "Your package does not have classifiers data.",

--- a/setup.cfg
+++ b/setup.cfg
@@ -10,13 +10,13 @@ classifiers =
     Operating System :: OS Independent
     Programming Language :: Python
     Programming Language :: Python :: 3
-    Programming Language :: Python :: 3.6
     Programming Language :: Python :: 3.7
     Programming Language :: Python :: 3.8
     Programming Language :: Python :: 3.9
     Programming Language :: Python :: 3.10
     Programming Language :: Python :: 3.11
     Programming Language :: Python :: 3 :: Only
+    Programming Language :: Python :: Implementation :: CPython
     Programming Language :: Python :: Implementation :: PyPy
 keywords = pypi, quality, testing
 author = Lennart Regebro
@@ -32,6 +32,7 @@ include_package_data = True
 packages = find:
 package_dir =
     = .
+python_requires = >=3.7
 install_requires =
     build
     docutils

--- a/setup.cfg
+++ b/setup.cfg
@@ -34,9 +34,8 @@ package_dir =
     = .
 python_requires = >=3.7
 install_requires =
-    build
+    build>=0.7.0
     docutils
-    pep517
     pygments
     requests
     setuptools>=61.0.0
@@ -51,7 +50,6 @@ test =
     pytest
     pytest-cov
     setuptools>=60
-    flit >=3.4,<4
     zest.releaser[recommended]
 
 [options.entry_points]

--- a/setup.cfg
+++ b/setup.cfg
@@ -38,7 +38,7 @@ install_requires =
     docutils
     pygments
     requests
-    setuptools>=61.0.0
+    setuptools>=42
     trove-classifiers>=2022.6.26
     wheel
 

--- a/setup.cfg
+++ b/setup.cfg
@@ -36,11 +36,11 @@ python_requires = >=3.7
 install_requires =
     build>=0.7.0
     docutils
+    packaging
     pygments
     requests
     setuptools>=42
     trove-classifiers>=2022.6.26
-    wheel
 
 [options.packages.find]
 where = .

--- a/setup.cfg
+++ b/setup.cfg
@@ -47,8 +47,6 @@ where = .
 
 [options.extras_require]
 test =
-    pytest
-    pytest-cov
     setuptools>=60
     zest.releaser[recommended]
 
@@ -69,7 +67,6 @@ testpaths =
 [check-manifest]
 ignore =
   .pre-commit-hooks.yaml
-  fetch_classifiers.py
 
 [zest.releaser]
 create-wheel = yes

